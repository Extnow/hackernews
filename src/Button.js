import React from 'react';

export default ({ onClick, className = '', children }) => (
  <button onClick={onClick} className={className} type="button">{children}</button>
);
