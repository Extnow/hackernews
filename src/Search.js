import React from 'react';

export default ({
  value, onChange, onSubmit, children,
}) => (
  <form onSubmit={onSubmit}>
    <input type="text" onChange={onChange} value={value} />
    <button type="submit">{children}</button>
  </form>
);
