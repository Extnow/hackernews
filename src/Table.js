import React from 'react';
import Button from './Button';

const largeColumn = {
  width: '40%',
};

const midColumn = {
  width: '30%',
};

const smallColumn = {
  width: '10%',
};

export default ({ articles, onDismiss }) => (
  <div className="table">
    {articles.map(item => (
      <div key={item.objectID} className="table-row">
        <span style={largeColumn}>
          <a href={item.url}>{item.title}</a>
        </span>
        <span style={midColumn}>{item.author}</span>
        <span style={smallColumn}>{item.num_comments}</span>
        <span style={smallColumn}>{item.points}</span>
        <span style={smallColumn}>
          <Button onClick={() => onDismiss(item.objectID)} className="button-inline">
            Отклонить
          </Button>
        </span>
      </div>
    ))}
  </div>
);
