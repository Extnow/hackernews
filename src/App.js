import React from 'react';
import Search from './Search';
import Table from './Table';
import Button from './Button';
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '5';

const PATH_BASE = 'https://hn.algoli1a.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

export default class App extends React.Component {
  state = {
    articles: null,
    searchTerm: DEFAULT_QUERY,
    error: null,
  };

  componentDidMount() {
    const { searchTerm } = this.state;

    this.fetchSearchTopStories(searchTerm);
  }

  fetchSearchTopStories = (searchTerm, page = 0) => {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(error => this.setState({ error }));
  };

  setSearchTopStories = (articles) => {
    const { hits, page } = articles;

    const oldHits = page !== 0 ? hits : [];

    const updateHits = [...oldHits, ...hits];

    this.setState({
      articles: { hits: updateHits, page },
    });
  };

  onSearchSubmit = (event) => {
    const { searchTerm } = this.state;

    this.fetchSearchTopStories(searchTerm);
    event.preventDefault();
  };

  onDismiss = (id) => {
    const { articles } = this.state;

    const isNotId = item => item.objectID !== id;
    const updateHits = articles.hits.filter(isNotId);
    this.setState({
      articles: { ...articles, hits: updateHits },
    });
  };

  onSearchChange = (event) => {
    this.setState({ searchTerm: event.target.value });
  };

  render() {
    const { articles, searchTerm, error } = this.state;
    const page = (articles && articles.page) || 0;

    if (error) {
      return <p>{error.message}</p>;
    }

    if (!articles) {
      return null;
    }

    return (
      <div className="page">
        <div className="interactions">
          <Search value={searchTerm} onSubmit={this.onSearchSubmit} onChange={this.onSearchChange}>
            Поиск
          </Search>
        </div>
        <Table articles={articles.hits} onDismiss={this.onDismiss} />
        <div className="interaction">
          <Button onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}>
            Больше историй
          </Button>
        </div>
      </div>
    );
  }
}
