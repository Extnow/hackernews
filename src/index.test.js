import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from './App';
import Search from './Search';
import Button from './Button';
import Table from './Table';

describe('App', () => {
  it('отрисовывает без ошибки', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('есть корректный снимок', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search', () => {
  it('отрисовывает без ошибки', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Поиск</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('есть корректный снимок', () => {
    const component = renderer.create(<Search>Поиск</Search>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  it('отрисовывает без ошибки', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Больше историй</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('есть корректный снимок', () => {
    const tree = renderer
      .create(<Button>Больше историй</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// describe('Table', () => {

//   const props = {
//     list: [
//       { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
//       { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
//     ],
//   };

//   it('отрисовывает без ошибки', () => {
//     const div = document.createElement('div');
//     ReactDOM.render(<Table { ...props } />, div);
//   });

//   test('есть корректный снимок', () => {
//     const component = renderer.create(<Table { ...props } />);
//     const tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   });
// });
