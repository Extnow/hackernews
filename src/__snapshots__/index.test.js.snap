// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App есть корректный снимок 1`] = `null`;

exports[`Button есть корректный снимок 1`] = `
<button
  className=""
  onClick={undefined}
  type="button"
>
  Больше историй
</button>
`;

exports[`Search есть корректный снимок 1`] = `
<form
  onSubmit={undefined}
>
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    type="submit"
  >
    Поиск
  </button>
</form>
`;
